apps/emqx/src/emqx_channel.erl

```diff
@@ -1220,14 +1221,18 @@ handle_info(
     {sock_closed, Reason},
     Channel =
         #channel{
-            conn_state = ConnState
+            conn_state = ConnState,
+            clientinfo = ClientInfo,
+            session = Session
         }
 ) when
     ConnState =:= connected orelse ConnState =:= reauthenticating
 ->
+    {Intent, Session1} = emqx_session:disconnect(ClientInfo, Session),
```

Comment:
Discussion point. @thalesmg @ieQu1 

Here we ask the session what it wants the channel to do upon disconnect (`Intent`), because we don't really want the channel backed by new persistent session implementation to live after the client has gone. This requires one more function in the session API surface, and I naturally wonder if there's a way to simplify this.

For reference, [this is](https://github.com/emqx/emqx/blob/b8184966ab40e847c2b0b5e2bf6399aa48fff0fc/apps/emqx/src/emqx_session_mem.erl#L713-L715) how `emqx_session_mem` implements it, and [this is](https://github.com/emqx/emqx/blob/b8184966ab40e847c2b0b5e2bf6399aa48fff0fc/apps/emqx/src/emqx_persistent_session_ds.erl#L283-L284) how `emqx_persistent_session_ds` does.

Reply:
> That is, there's no need to keep the channel process around for takeover when we have persistent sessions, so we terminate it?

Right. Also parts of `emqx_channel` logic that kick in between channel disconnect and expiration / takeover are specific to `emqx_session_mem`. So this is also a kind of safety measure, to ensure that idling channel with ds persistent sessions processes will not crash when they receive some message for some reason.

<sup>Thanks for the question! It seems that I missed the fact that a channel with in-memory session can still try to takeover a persistent session (which will fail miserably). I'll need to take that into account, however it sounds like a stuff for a followup PRs, especially given that the semantics of that case are unclear.</sup>